src/RedCode/Currency/Rate/Provider/YahooCurrencyRateProvider.php

```diff
@@ -0,0 +1,158 @@
+<?php
+
+namespace RedCode\Currency\Rate\Provider;
+
+use RedCode\Currency\ICurrency;
+use RedCode\Currency\ICurrencyManager;
+use RedCode\Currency\Rate\ICurrencyRate;
+use RedCode\Currency\Rate\ICurrencyRateManager;
+
+class YahooCurrencyRateProvider implements ICurrencyRateProvider
+{
+    const PROVIDER_NAME = 'yahoo';
+
+    public function __construct(ICurrencyRateManager $currencyRateManager, ICurrencyManager $currencyManager)
+    {
+        $this->currencyRateManager  = $currencyRateManager;
+        $this->currencyManager      = $currencyManager;
+    }
+
+    /**
+     * Load rates by date
+     *
+     * @param ICurrency[] $currencies
+     * @param \DateTime|null $date
+     *
+     * @return ICurrencyRate[]
+     */
+    public function getRates($currencies, \DateTime $date = null)
+    {
+        $currencyCodes = [];
+
+        foreach ($currencies as $currency) {
+            $currencyCodes[] = $currency->getCode();
+        }
+
+        if (null === $date) {
+            $date = new \DateTime();
+            $results = $this->_processActualRequest($currencyCodes);
+        } else {
+            $results = $this->_processHistoricalRequest($currencyCodes, $date);
+        }
+
+        $rates = [];
+        foreach ($results as $code => $rate) {
+            $rates[$code] = $this->currencyRateManager->getNewInstance(
+                $this->currencyManager->getCurrency($code),
+                $this,
+                $date,
+                $rate,
+                1
+            );
+        }
+
+        return $rates;
+    }
+
+    /**
+     * @param $currencyCodes
+     *
+     * @return array
+     */
+    private function _processActualRequest($currencyCodes)
+    {
+        $baseUrl =  'http://query.yahooapis.com';
```

Comment:
It's better to build url params with [http_build_query](http://php.net/manual/en/function.http-build-query.php)